tools/isobuild/js-analyze.js

```diff
@@ -13,6 +19,112 @@ function tryToParse(source) {
   }
 }
 
+var dependencyKeywordPattern = /\b(require|import|export)\b/g;
+
+export function findImportedModuleIdentifiers(source) {
+  const identifiers = {};
+  const possibleIndexes = [];
+  let match;
+
+  dependencyKeywordPattern.lastIndex = 0;
+  while ((match = dependencyKeywordPattern.exec(source))) {
+    possibleIndexes.push(match.index);
+  }
+
+  if (!possibleIndexes.length) {
+    return {};
+  }
+
+  const ast = tryToParse(source);
+
+  function walk(node, left, right) {
```

Comment:
The basic idea is that `findImportedModuleIdentifiers` uses a regular expression to scan quickly for possible locations of certain tokens (`require`, `import`, `export`), then uses that location information to steer the AST traversal, so that it visits on subtrees that contain interesting tokens. The AST traversal determines if the tokens were actually what we thought they were (a `require` function call, or an `import` or `export` statement).